src/v/security/gssapi_principal_mapper.cc

```diff
@@ -0,0 +1,482 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "security/gssapi_principal_mapper.h"
+
+#include "security/logger.h"
+#include "vassert.h"
+#include "vlog.h"
+
+#include <boost/algorithm/string/case_conv.hpp>
+#include <fmt/format.h>
+#include <fmt/ostream.h>
+
+#include <charconv>
+
+namespace security {
+static constexpr std::string_view gssapi_name_pattern
+  = R"(([^/@]*)(/([^/@]*))?@([^/@]*))";
+static constexpr std::string_view non_simple_pattern = R"([/@])";
+static constexpr std::string_view parameter_pattern = R"(([^$]*)(\$(\d*))?)";
+static constexpr std::string_view rule_pattern
+  = R"(((DEFAULT)|((RULE:\[(\d*):([^\]]*)](\(([^)]*)\))?(s\/([^\/]*)\/([^\/]*)\/(g)?)?\/?(L|U)?))))";
+
+namespace detail {
+std::vector<gssapi_rule> parse_rules(
+  std::string_view default_realm,
+  std::optional<std::vector<ss::sstring>> unparsed_rules) {
+    static thread_local const re2::RE2 rule_parser(
+      rule_pattern, re2::RE2::Quiet);
+
+    vassert(
+      rule_parser.ok(),
+      "Failed to build rule pattern regex: {}",
+      rule_parser.error());
+
+    if (!unparsed_rules || unparsed_rules.value().empty()) {
+        return {gssapi_rule(default_realm)};
+    }
+
+    std::vector<gssapi_rule> rv;
+    re2::StringPiece default_;
+    re2::StringPiece num_components_str;
+    re2::StringPiece format;
+    re2::StringPiece match_regex;
+    re2::StringPiece from_pattern;
+    re2::StringPiece to_pattern;
+    re2::StringPiece repeat;
+    re2::StringPiece upper_lower;
+    for (const auto& rule : *unparsed_rules) {
+        const re2::StringPiece rule_piece(rule.data(), rule.size());
+        if (!re2::RE2::FullMatch(
+              rule_piece,
+              rule_parser,
+              nullptr,
+              &default_,
+              nullptr,
+              nullptr,
+              &num_components_str,
+              &format,
+              nullptr,
+              &match_regex,
+              nullptr,
+              &from_pattern,
+              &to_pattern,
+              &repeat,
+              &upper_lower)) {
+            throw std::runtime_error("GSSAPI: Invalid rule: " + rule);
```

Comment:
this needs a prometheus metric.

Reply:
@michael-redpanda i think the metrics can be collected at the kafka layer as these errors bubble up to the connection level?

Reply:
This logic verifies that the rule provided by the _user_ is good and is not when the rule is actually applied to the kerberos principal.

As in, checks that "RULE:[1:$1](redpanda.*)s/(redpanda).*/$1rocks/g" is good.

Reply:
> @michael-redpanda i think the metrics can be collected at the kafka layer as these errors bubble up to the connection level?

I don't know :)  Would love to have a conversation about this to better understand what you're looking for